src/borg/repository.py

```diff
@@ -750,10 +750,66 @@ def __contains__(self, id):
         return id in self.index
 
     def list(self, limit=None, marker=None):
+        """
+        list <limit> IDs starting from after id <marker> - in index (pseudo-random) order.
+        """
         if not self.index:
             self.index = self.open_index(self.get_transaction_id())
         return [id_ for id_, _ in islice(self.index.iteritems(marker=marker), limit)]
 
+    def scan(self, limit=None, marker=None):
+        """
+        list <limit> IDs starting from after id <marker> - in on-disk order, so that a client
+        fetching data in this order does linear reads and reuses stuff from disk cache.
+        """
+        if limit is not None and limit < 1:
+            raise ValueError('please use limit > 0 or limit = None')
+        if not self.index:
+            transaction_id = self.get_transaction_id()
+            self.index = self.open_index(transaction_id)
+        # create a second index object - use it to discover object IDs we can not discover on-disk
+        # (either because they are not there, just in the index, or because they have IntegrityErrors).
```

Comment:
Yes, the description wasn't very accurate.

The second index was also (hidden) state, in which case the counter would work. It could also be made explicit state, by making `marker` a tuple (object-id, object-counter) which is returned explicitly instead of implicitly (objects_ids[-1]), or similar. Ofc. with the current client code there is no problem doing this in the client, as it already has a complete copy of the repository index obtained at that point. So as it is now (a3cd5f3) would work and is arguably the simplest/neatest solution.